### PR TITLE
THRIFT-4489: Add UDS support for nodejs thrift client

### PR DESCRIPTION
**Pull request by danielhtshih**

### CODE REVIEW

**Comment by jeking3** [comment]
Where does the "--type uds" option get passed into the python test client or server in order to convince it to use domain sockets?  In other test suites like cpp, only "--domain_socket" is required in order to force it to use "domain".

File: test/tests.json [hunk 1]
```diff
@@ -221,7 +221,8 @@
       "framed"
     ],
     "sockets": [
-      "ip"
+      "ip",
+      "domain"
```
Comment (jeking3) [inline test/tests.json]: You added domain tests to the haskell group, not the nodejs group.

**Comment by jeking3** [comment]
The presence of "--domain-socket" to the client or server implies uds.  This needs to be fixed, and domain needs to be moved from haskell to nodejs, and the cross tests that are failing (many) in the ubuntu-artful docker image need to be investigated.  For example it is likely expected for the "http-domain" subset of tests to fail given the http-ip ones fail already, however in one test I found that the call to fs.unlinkSync(path)  [server.js] failed because the domain socket doesn't exist before the test begins, and this call throws an exception, causing it to fail.

**Comment by jeking3** [comment]
If you haven't seen it I recommend reading `build/docker/README.md` and use the `ubuntu-artful` image to run a cross test with `build/docker/scripts/cross-test.sh` inside the docker container.  Doing this effectively runs the same build as travis would (job #1 which is the cross test).  You can run only the nodejs tests with domain sockets by running the command `test/test.py --regex '.*nodejs.*domain.*'` after completing cross-test.sh.

**Comment by jeking3** [comment]
Please rebase on master one more time, most of the build issues are gone now.

**Comment by danielhtshih** [comment]
I believe the unexpected failures from the latest run is not related to my changes for nodejs.

Maybe it needs another merge from master branch? 
```
===============================================================================
Following 14 tests were expected to cleanly succeed but needed retry:
===============================================================================
server-client:          protocol:         transport:               result:
nodejs-nodejs           json              buffered-domain          flaky(1 retry)
nodejs-cpp              json              buffered-ip-ssl          flaky(3 retries)
hs-csharp               json              framed-ip                flaky(2 retries)
cpp-cpp                 multic            framed-ip-ssl            flaky(2 retries)
cpp-cpp                 multij-json       http-ip-ssl              flaky(1 retry)
cpp-cpp                 multij-json       framed-ip-ssl            flaky(1 retry)
cpp-cpp                 multih            http-ip-ssl              flaky(1 retry)
cpp-cpp                 multih-header     http-ip-ssl              flaky(2 retries)
cpp-cpp                 multih-header     buffered-ip-ssl          flaky(2 retries)
cpp-cpp                 multic-compact    http-ip-ssl              flaky(1 retry)
cpp-cpp                 multic-compact    buffered-ip-ssl          flaky(3 retries)
cpp-cpp                 multij-json       http-ip-ssl              flaky(1 retry)
cpp-cpp                 multij-json       buffered-ip-ssl          flaky(1 retry)
cpp-cpp                 multij            buffered-ip-ssl          flaky(2 retries)
===============================================================================
*** Following 1 failures were unexpected ***:
If it is introduced by you, please fix it before submitting the code.
===============================================================================
server-client:          protocol:         transport:               result:
cpp-cpp                 multij-json       buffered-ip-ssl          failure(64)
===============================================================================
```

**Comment by jeking3** [comment]
Perhaps; my PR for THRIFT-4515 improves the stability and coverage of crosstest by adding more controlled server shutdown, and it passed tests (I've run it many times on my local fork as well).  I just merged it into master, so if you want to squash to a single commit and rebase, let's give it a try.

**Comment by jeking3** [comment]
I am running a cross test on this locally; if it passes I will merge it, so don't take any action.

**Comment by jeking3** [comment]
I get an error that may be unrelated when I run cross tests on --server nodejs locally.  This one test failes sporadically and I don't know if it could be related to these changes; the client has to be killed in this scenario, I don't know if it is because the server is unresponsive or not.  I may just disable this one test for the merge.

```
===============================================================================
*** Following 1 failures were unexpected ***:
If it is introduced by you, please fix it before submitting the code.
===============================================================================
server-client:          protocol:         transport:               result:
nodejs-cpp              json              buffered-ip-ssl          failure(timeout)
===============================================================================
Unexpected failures are logged to test/log/unexpected_failures.log
You can browse results at:
	file:///thrift/src/test/index.html
# If you use Chrome, run:
# 	cd /thrift/src
#	python -m http.server 8001
# then browse:
# 	http://localhost:8001/test/
Full log for each test is here:
	test/log/server_client_protocol_transport_client.log
	test/log/server_client_protocol_transport_server.log
1 failed of 193 tests in total.
Test execution took 30.1 seconds.
Mon Mar 19 19:40:15 2018
root@3d39d5d9eb3c:/thrift/src# tail test/log/nodejs-cpp_json_buffered-ip-ssl_client.log 
testBinary(siz = 4096)
testBinary(siz = 8192)
testBinary(siz = 16384)
testBinary(siz = 32768)
testBinary(siz = 65536)
testBinary(siz = 131072)
===============================================================================
Return code: -15 (negative values indicate kill by signal)
Test execution took 8.0 seconds.
Mon Mar 19 19:40:11 2018
root@3d39d5d9eb3c:/thrift/src# tail test/log/nodejs-cpp_json_buffered-ip-ssl_server.log 
Mon Mar 19 19:40:03 2018
Executing: node server.js --type=tcp --protocol=json --transport=buffered --ssl --port=39155
Directory: /thrift/src/lib/nodejs/test
config:delay: 5
config:timeout: 8
===============================================================================
===============================================================================
Return code: -1 (negative values indicate kill by signal)
Test execution took 8.3 seconds.
Mon Mar 19 19:40:11 2018
```

**Comment by jeking3** [comment]
Okay so, here's the deal.  This pull request has 9 commits and one of them is a merge.  If you want your name on the commit tag you will need to squash and resubmit with a single commit and then I can cherry-pick it over and push it.  Otherwise if you don't care, I can squash and merge it, but my name will end up in the "Author" field of the commit, and I will include your name in the content of the pull request, i.e.:

```
commit 496568e57a17889f484c6275078bef2bcb43382e (HEAD -> master)
Author: James E. King III <jking@apache.org>
Date:   Mon Mar 19 16:16:37 2018 -0400

    THRIFT-4489: Add UDS support for nodejs thrift client
    Patch: Daniel Shih <hotingwow@gmail.com>
    Client: nodejs
    
    This closes #1491
```
Let me know what you want to do.  I'm fine either way.

**Comment by danielhtshih** [comment]
@jeking3 please use this single commit https://github.com/apache/thrift/pull/1491/commits/041f91dd6042b98c5bbd0846f5e6ed7fb6fd4bf1

Thanks.

**Comment by jeking3** [comment]
This squash was not done properly.  It includes all of the master commits in it.

You want to start with an updated master:

git checkout master
git pull

Then you want to branch:
git checkout -b THRIFT-4489-squashed

Then you want to merge with squash, pulling anything in uds-nodejs over to THRIFT-4489-squashed
git merge --squash uds-nodejs

Then commit and add your description for your changes.
Then push that branch and let me know when it is in your fork, and I will pull it and merge it.  No need to update this PR or open a new one.

**Comment by danielhtshih** [comment]
@jeking3 here is the single commit https://github.com/danielhtshih/thrift/commit/9c560daf52fcd89c212451f0793973af17f6cc55 from THRIFT-4489-squashed

Thanks.

**Comment by jeking3** [comment]
Looks good, will merge in the morning.

**Pull request closed without merging**